more_itertools/more.py

```diff
@@ -694,43 +694,28 @@ def windowed(seq, n, fillvalue=None, step=1):
     to the left:
 
         >>> iterable = [1, 2, 3, 4]
-        >>> n = 3
-        >>> padding = [None] * (n - 1)
-        >>> list(windowed(chain(padding, iterable), 3))
+        >>> width = 3
+        >>> padding = [None] * (width - 1)
+        >>> list(windowed(chain(padding, iterable), width))
         [(None, None, 1), (None, 1, 2), (1, 2, 3), (2, 3, 4)]
-
     """
-    if n < 0:
-        raise ValueError('n must be >= 0')
-    if n == 0:
-        yield tuple()
-        return
-    if step < 1:
-        raise ValueError('step must be >= 1')
 
-    it = iter(seq)
-    window = deque([], n)
-    append = window.append
-
-    # Initial deque fill
-    for _ in range(n):
-        append(next(it, fillvalue))
-    yield tuple(window)
-
-    # Appending new items to the right causes old items to fall off the left
-    i = 0
-    for item in it:
-        append(item)
-        i = (i + 1) % step
-        if i % step == 0:
-            yield tuple(window)
-
-    # If there are items from the iterable in the window, pad with the given
-    # value and emit them.
-    if (i % step) and (step - i < n):
-        for _ in range(step - i):
-            append(fillvalue)
-        yield tuple(window)
+    def helper(**kwargs):
+        window, i = deque((), width), width
+        for _ in map(window.append, iterable):
+            i -= 1
+            if not i:
+                i = step
+                yield window
+        if 0 < len(window) < width:
+            yield chain(window, (fillvalue,) * (width - len(window)))
+        elif 0 < i < min(step, width):
+            window.extend((fillvalue,) * i)
+            yield window
+
+    if width >= 0 < step:
+        return map(tuple, helper(**locals()) if width else ((),))
```

Comment:
Is it really necessary to inline `locals` here? What's the performance impact?

Reply:
This is my jamb, I was tired there ) Only **step** is critical, so changing the signature of the helper to a **helper(step)** will be right.